File: programmers/dp/숫자-타자-대회.py
def solution(numbers):
    positions = {
        1: (0, 0),
        2: (0, 1),
        3: (0, 2),
        4: (1, 0),
        5: (1, 1),
        6: (1, 2),
        7: (2, 0),
        8: (2, 1),
        9: (2, 2),
        0: (3, 1)
    }
    
    
    def get_cost(a, b):
        if a == b:
            return 1
        xa, ya = positions[a]
        xb, yb = positions[b]
        
        dx, dy = abs(xa - xb), abs(ya - yb)
        p, q = min(dx, dy), max(dx, dy)
        return  2 * (q - p) + 3 * p
    
    
    # dp[i][l][r] : i번째 버튼을 누를 때, 오른손이 r, 왼손이 l 위치에 있을 때의 소요 시간?
    
#     q = deque([(0, 4, 6)])
#     map = dict()
    
#     for index in range(n):
        
#         number = int(numbers[index])
        
#         for _ in range(len(q)):
#             cost, left, right = q.popleft()
#             q.append((cost + get_cost(left, number), number, right))
#             q.append((cost + get_cost(right, number), left, number))
                     
#     # print(q)
#     return min([cost for cost, _, _ in q])
    
    ### DP
#     dp = [[[float("inf") for _ in range(10)] for _ in range(10)] for _ in range(n + 1)]
#     dp[0][4][6] = 0
    
#     for i in range(1, n + 1):
#         cur = int(numbers[i - 1])
#         for r in range(10):
#             for l in range(10):
#                 dp[i][cur][l] = min(dp[i][cur][l], dp[i - 1][r][l] + cost(r, cur))
#                 dp[i][r][cur] = min(dp[i][r][cur], dp[i - 1][r][l] + cost(l, cur))
    
    
#     res = float("inf")
#     for r in range(10):
#         for l in range(10):
#             res = min(res, dp[-1][r][l])

    # (left, right) -> min_cost
    dp = {(4, 6): 0}

    for ch in numbers:
        num = int(ch)
        next_dp = {}

        for (left, right), cost in dp.items():
            # 왼손으로 누르기
            if num != right:
                new_cost = cost + get_cost(left, num)
                state = (num, right)
                if state not in next_dp or next_dp[state] > new_cost:
                    next_dp[state] = new_cost

            # 오른손으로 누르기
            if num != left:
                new_cost = cost + get_cost(right, num)
                state = (left, num)
                if state not in next_dp or next_dp[state] > new_cost:
                    next_dp[state] = new_cost

        dp = next_dp

    res = min(dp.values())
    return res

File: programmers/dp/숫자-타자-대회.cpp
#include <string>
#include <vector>

using namespace std;

int INF = 1e9;
int X[] = {
    1, // 0
    0, 1, 2,
    0, 1, 2,
    0, 1, 2,
};

int Y[] = {
    3, // 0
    0, 0, 0,
    1, 1, 1,
    2, 2, 2,
};

int cost(int a, int b) {
    int dx = abs(X[a] - X[b]);
    int dy = abs(Y[a] - Y[b]);
    if (dx == 0 && dy == 0)
        return 1;
    return min(dx, dy) * 3 + (max(dx, dy) - min(dx, dy)) * 2;
}

int solution(string numbers) {
    
    vector<vector<int>> dp(10, vector<int>(10, INF));
    dp[4][6] = 0;
    // dp[left][right] -> minimum cost;
    
    for (int i = 0; i < numbers.size(); i++) {
        
        int cur = numbers[i] - '0';
        vector<vector<int>> new_dp(10, vector<int>(10, INF));
        
        for (int left = 0; left < 10; left++) {
            for (int right = 0; right < 10; right++) {    
                if (cur == left) {
                    new_dp[cur][right] = min(new_dp[cur][right], cost(left, cur) + dp[left][right]);
                } else if (cur == right) {
                    new_dp[left][cur] = min(new_dp[left][cur], cost(right, cur) + dp[left][right]);   
                } else {
                    new_dp[cur][right] = min(new_dp[cur][right], cost(left, cur) + dp[left][right]);
                    new_dp[left][cur] = min(new_dp[left][cur], cost(right, cur) + dp[left][right]);   

                }
            }
        }
        dp = new_dp;
    }
    
    int res = INF;
    for (int i = 0; i < 10; i++)
        for (int j = 0; j < 10; j++)
            res = min(res, dp[i][j]);
    
    return res;
}
